File: tests/typesafety/python/api_dict.py
# expect: passes
# title: The handler checks the order's shape by hand
import json


def price_cents(order: dict[str, object]) -> int:
    quantity = order.get("quantity")
    if not isinstance(quantity, int):
        raise ValueError("quantity missing or not a number")
    return quantity * 250


def handle(body: str) -> int:
    order = json.loads(body)
    if not isinstance(order, dict):
        raise ValueError("not an object")
    return price_cents(order)

File: tests/typesafety/typescript/api_dict.ts
// expect: passes

function priceCents(order: Record<string, unknown>): number {
  const quantity = order["quantity"];
  if (typeof quantity !== "number") {
    throw new Error("quantity missing or not a number");
  }
  return quantity * 250;
}

export function handle(body: string): number {
  const order: unknown = JSON.parse(body);
  if (typeof order !== "object" || order === null) {
    throw new Error("not an object");
  }
  return priceCents(order as Record<string, unknown>);
}
